### PR TITLE
Downgrade V4 roles to V3 at webapi endpoints

### PR DESCRIPTION
**Pull request by nklaassen**
Follow-up to https://github.com/gravitational/teleport/pull/7118.

I've realized that older teleport versions (including 5.2.1) will fetch roles from the apiserver endpoints, rather than the GRPC endpoint. This PR downgrades roles to V3 at the apiserver endpoint so that older teleport versions will always get V3 roles.

It looks like the apiserver role endpoints are no longer used in up-to-date versions of teleport and are scheduled to be removed in 7.0, so it seems correct to unconditionally downgrade the roles here.

I have tested this with a dual-cluster setup and a full upgrade process (seperate auth, proxy, and nodes on main and remote clusters).

I have already integrated this change to the backports for the original V4 Roles PR, https://github.com/gravitational/teleport/pull/7270 and https://github.com/gravitational/teleport/pull/7269.

### CODE REVIEW

File: lib/auth/apiserver.go [hunk 1]
```diff
@@ -2145,7 +2145,15 @@ func (s *APIServer) getRole(auth ClientI, w http.ResponseWriter, r *http.Request
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	return rawMessage(services.MarshalRole(role, services.WithVersion(version), services.PreserveResourceID()))
+	roleV4, ok := role.(*types.RoleV4)
+	if !ok {
+		return nil, trace.BadParameter("Unrecognized role version")
```
Comment (russjones) [inline lib/auth/apiserver.go:2150]: Lowercase error messages: `unrecognized role version`.

File: lib/auth/apiserver.go [hunk 2]
```diff
@@ -2145,7 +2145,15 @@ func (s *APIServer) getRole(auth ClientI, w http.ResponseWriter, r *http.Request
...
+	}
+	downgraded, err := downgradeRole(context.Background(), roleV4)
```
Comment (russjones) [inline lib/auth/apiserver.go:2152]: Can you add a comment block here with what version this migration code can be removed? Search for `DELETE IN` comments in the code base and follow that pattern. Will make it easy to remove migration logic in the future.

**Review by russjones — APPROVED** [review]
Looks good to me, but let's wait for another reviewer as well.

File: lib/auth/apiserver.go (hunk 2 above)
Comment (nklaassen) [reply lib/auth/apiserver.go:2154]: All of the apiserver role endpoints have `DELETE IN 7.0` where they are registered, I've added comments at the function definitions as well https://github.com/gravitational/teleport/blob/6684b13672cc5706d687bb9001b40e158476f447/lib/auth/apiserver.go#L194-L199

File: lib/auth/apiserver.go [hunk 3]
```diff
@@ -2140,22 +2141,40 @@ func (s *APIServer) upsertRole(auth ClientI, w http.ResponseWriter, r *http.Requ
...
 		return nil, trace.Wrap(err)
 	}
 	out := make([]json.RawMessage, len(roles))
 	for i, role := range roles {
-		raw, err := services.MarshalRole(role, services.WithVersion(version), services.PreserveResourceID())
+		roleV4, ok := role.(*types.RoleV4)
+		if !ok {
+			return nil, trace.BadParameter("unrecognized role version")
+		}
+		downgraded, err := downgradeRole(context.Background(), roleV4)
```
Comment (smallinsky) [inline lib/auth/apiserver.go:2173]: nit: just for consistency 
```suggestion
		downgraded, err := downgradeRole(r.Context(), roleV4)
```

**Review by smallinsky — APPROVED** [review]

**Review by russjones — APPROVED** [review]
Bot.

File: lib/auth/apiserver.go (hunk 3 above)
Comment (nklaassen) [reply lib/auth/apiserver.go]: Good catch, thanks

**Pull request merged**